pkg/api/licenses.go

```diff
@@ -884,3 +885,306 @@ func SearchInLicense(c *gin.Context) {
 	}
 	c.JSON(http.StatusOK, res)
 }
+
+// ImportLicenses creates new licenses records via a json file.
+//
+//	@Summary		Import licenses by uploading a json file
+//	@Description	Import licenses by uploading a json file
+//	@Id				ImportLicenses
+//	@Tags			Licenses
+//	@Accept			multipart/form-data
+//	@Produce		json
+//	@Param			file	formData	file true "licenses json file list"
+//	@Success		200		{object}	models.ImportLicensesResponse{data=[]models.LicenseImportStatus}
+//	@Failure		400		{object}	models.LicenseError	"input file must be present"
+//	@Failure		500		{object}	models.LicenseError	"Internal server error"
+//	@Security		ApiKeyAuth
+//	@Router			/licenses/import [post]
+func ImportLicenses(c *gin.Context) {
+	username := c.GetString("username")
+	file, header, err := c.Request.FormFile("file")
+	if err != nil {
+		er := models.LicenseError{
+			Status:    http.StatusBadRequest,
+			Message:   "input file must be present",
+			Error:     err.Error(),
+			Path:      c.Request.URL.Path,
+			Timestamp: time.Now().Format(time.RFC3339),
+		}
+		c.JSON(http.StatusBadRequest, er)
+		return
+	}
+	defer file.Close()
+
+	if filepath.Ext(header.Filename) != ".json" {
+		er := models.LicenseError{
+			Status:    http.StatusBadRequest,
+			Message:   "only files with format *.json are allowed",
+			Error:     "only files with format *.json are allowed",
+			Path:      c.Request.URL.Path,
+			Timestamp: time.Now().Format(time.RFC3339),
+		}
+		c.JSON(http.StatusBadRequest, er)
+		return
+	}
+
+	var licenses []models.LicenseImport
+	decoder := json.NewDecoder(file)
+	if err := decoder.Decode(&licenses); err != nil {
+		er := models.LicenseError{
+			Status:    http.StatusInternalServerError,
+			Message:   "invalid json",
+			Error:     err.Error(),
+			Path:      c.Request.URL.Path,
+			Timestamp: time.Now().Format(time.RFC3339),
+		}
+		c.JSON(http.StatusInternalServerError, er)
+		return
+	}
+
+	res := models.ImportLicensesResponse{
+		Status: http.StatusOK,
+	}
+
+	for _, license := range licenses {
+		_ = db.DB.Transaction(func(tx *gorm.DB) error {
+			newLicenseMap := make(map[string]interface{})
+			if license.Shortname.IsDefinedAndNotNull {
+				newLicenseMap["rf_shortname"] = license.Shortname.Value
+			} else {
+				res.Data = append(res.Data, models.LicenseError{
+					Status:    http.StatusInternalServerError,
+					Message:   "field rf_shortname cannot be null",
+					Error:     "",
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				})
+				return errors.New("field rf_shortname cannot be null")
```

Comment:
The repeated creation of error data can be converted into a small function with takes in the `res.Data` and field name and returns the updated `res.Data` and error object. This can reduce the size a lot.